### PR TITLE
Adding DotNotation Support and Widget Framework to existing cli (#1495)

### PR DESCRIPTION
**Pull request by LeighFinegold**
## Overview

This PR introduces the foundational **Widget Framework** and **Dot Notation** capabilities for CALM templating, putting them structurally in the right place:
- **Dot Notation**: Available for both `calm template` and `calm docify` commands
- **Widget Framework**: Specifically designed for `calm docify` command

Related to: [CALM Docify CLI v1.0.0 MVP #1452](https://github.com/finos/architecture-as-code/issues/1452)

### Dot Notation Support (for calm template & calm docify)
- ✅ Template preprocessor for natural path syntax (`nodes['unique-id=="service"']`)
- ✅ Support for bracket notation, filters, and mixed expressions
- ✅ Handlebars keyword protection

### Widget Framework (for calm docify only)
- ✅ Standalone widget system with registry, engine, and renderer
- ✅ Type-safe widget validation and error handling  
- ✅ Three foundational widgets: `table`, `list`, `json-viewer`

### Template Engine Integration
- ✅ Naturally integrates with new canonical model and dereferencing (#1471)
- ✅ Path extraction and processing pipeline

### Dot Notation (Both Commands)
```handlebars
{{!-- Natural dot notation --}}
{{ architecture.nodes['service-unique-id'].description }}

{{!-- Advanced filtering --}}
{{ nodes['node-type=="service"'] }}

{{!-- Mixed notation --}}
{{ relationships['type=="connect"'].nodes }}
```

### Widgets (calm docify Only)
```handlebars
{{!-- Widget rendering with dot notation paths --}}
{{table nodes['node-type=="service"'] columns="name,port"}}
{{list relationships['type=="connect"'].nodes}}
{{json-viewer architecture.metadata}}
```

### Widgets to be Enhanced
- Expand validation and options for existing widgets

### Dot Notation to be Enhanced  
- Extended capabilities leveraging JSONPath internally
- Cleanup of core code with better logging and aid to end users

### CALM-Specific Widgets (Next Phase)
- `flow-sequence` - Interactive sequence diagrams
- `c4-architecture` - System architecture views  
- `related-nodes` - Node relationship visualization
- Cleanup of current template bundle for docify website mode

## Known TODOs

The following has been identified but should not block this initial PR. This will allow other maintainers to dive in and contribute. 

- **Helper Function Conflicts**: Resolve naming between widget helpers and the default transformer --tempalte and template-dir are using under covers.
- **Architecture Refactoring**: Move docifier/graphing packages to widget framework  
- **Error Message Enhancement**: More detailed widget validation feedback
- **Handlebars Isolation**: Dependency injection for better testability
- **Documentation**: Documentation to be added to main docs website prior for v1.0 release.  

## Testing

- ✅ Widget validation and rendering tests
- ✅ Integration tests with real CALM data
- ✅ Error handling and edge case coverage

## Migration

- ✅ Backward compatible with existing `calm template` and `calm docify`  workflows
- ✅ No breaking changes to current template bundles
- ✅ Gradual adoption path for new features

(@markscott-ms, @aidanm3341 , @willosborne , @Thels , @rocketstack-matt )

### CODE REVIEW

**Review by rocketstack-matt — DISMISSED** [review]
I've tested this locally and it is frankly incredibly cool . . . 

```md
# Example table
{{table nodes['load-balancer']}}

# Example list
{{list relationships property="unique-id"}}

# Example JSON View
{{json-viewer nodes}}
```

produces

Screenshot 2025-08-07 at 09 27 46

One big issue I came across we should fix was when initially passing the output file as in the same directory I was working in (which happened to be the root of the repo) . . . it deleted everything in the directory without warning 😢 

I think this is existing behavior, but I wasn't expecting it because I gave an output file not just a directory. Either way, I think we should at the very least add a warning that the output dir will be cleaned and prompt for confirmation. In the case of passing an output file (as opposed to a directory) I don't think it should clean out the directory at all.

File: calm-widgets/src/widget-engine.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    registerWidgetHelper(widgetId: string) {
+        this.handlebars.registerHelper(widgetId, (context: unknown, options: Record<string, unknown>) => {
+            const renderer = new WidgetRenderer(this.handlebars, this.registry);
+            const rendered = renderer.render(widgetId, context, options);
+            return new this.handlebars.SafeString(rendered);
+        });
+    }
+
+    registerDefaultWidgets() {
+        const widgets: { widget: CalmWidget<unknown, object, unknown>, folder: string }[] = [
```
Comment (markscott-ms) [inline calm-widgets/src/widget-engine.ts:50]: nitpick: have this list in alphabetical order of widget name?

File: calm-widgets/src/widget-helpers.spec.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
+import { describe, it, expect } from 'vitest';
+import { registerGlobalTemplateHelpers } from './widget-helpers';
+
+describe('Widget Helpers', () => {
+    const helpers = registerGlobalTemplateHelpers();
+
+    describe('eq helper', () => {
```
Comment (markscott-ms) [inline calm-widgets/src/widget-helpers.spec.ts:7]: Are tests for unequal types relevant?

`expect(helpers.eq(true, 'true').toBe(false)`

File: calm-widgets/src/widget-helpers.spec.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            expect(helpers.kebabToTitleCase(123)).toBe('');
+        });
+
+        it('handles already capitalized words', () => {
+            expect(helpers.kebabToTitleCase('API-Gateway')).toBe('API Gateway');
+        });
+    });
+
+    describe('kebabCase helper', () => {
+        it('converts strings to kebab-case', () => {
```
Comment (markscott-ms) [inline calm-widgets/src/widget-helpers.spec.ts:130]: Does it handle already kebabed strings?

`expect(helpers.kebabCase('API-GateWay')).toBe('api-gateway');`

File: calm-widgets/src/widget-registry.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import fs from 'fs';
+import path from 'path';
+import Handlebars from 'handlebars';
+import { CalmWidget} from './types';
```
Comment (markscott-ms) [inline calm-widgets/src/widget-registry.ts]: nitpick: inconsistent spacing within braces

**Comment by LeighFinegold** [comment]
> @LeighFinegold could you add issues for your known todos so they are more visible for others to assign and pick up?

Follow up issues will be raised over the next few days and will mark related items as good first issues for cli developers

File: calm-widgets/src/widgets/list/list-template.html [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{{#if ordered}}
```
Comment (markscott-ms) [inline calm-widgets/src/widgets/list/list-template.html:1]: There is a lot of duplicate code in this file - the only things that change are `ol` <=> `ul`. Can/should the `ordered` if/else be around the opening and closing tags only?

File: cli/src/cli.e2e.spec.ts [hunk 6]
```diff
@@ -50,7 +49,7 @@ describe('CLI Integration Tests', () => {
 
     afterAll(() => {
         if (tempDir) {
-            fs.rmSync(tempDir, { recursive: true, force: true });
+            //rs.rmSync(tempDir, { recursive: true, force: true });
```
Comment (markscott-ms) [inline cli/src/cli.e2e.spec.ts]: Is this line of code meant to be commented out?

**Review by markscott-ms — CHANGES_REQUESTED** [review]

File: cli/src/cli.e2e.spec.ts (hunk 6 above)
Comment (LeighFinegold) [reply cli/src/cli.e2e.spec.ts]: It has no impact to the build as ultimately everything is discarded in the end, but this is something that is done to view the outputs when there is an issue in more detail. It occurs on the cli.e2e.spec.ts tests because these are the tests where we on doing a full simulation of using the npm module and therefore harder to debug through problems i.e. this is done so i can debug using the index.js.  

Testing approach has been brought up in a few office hours. Will follow up with @willosborne and see if we can raise an epic around testing practices we are looking to adopt and cleanup.

File: calm-widgets/src/widget-engine.ts (hunk 1 above)
Comment (LeighFinegold) [reply calm-widgets/src/widget-engine.ts:50]: Raised https://github.com/finos/architecture-as-code/issues/1501

File: calm-widgets/src/widget-helpers.spec.ts (hunk 2 above)
Comment (LeighFinegold) [reply calm-widgets/src/widget-helpers.spec.ts:7]: That would fail the strict equals currently implemented. If we want to support those being the same then we should raise an issue.

File: calm-widgets/src/widgets/list/list-template.html (hunk 5 above)
Comment (LeighFinegold) [reply calm-widgets/src/widgets/list/list-template.html:1]: We could delegate the tag to the viewmodel and just print that. Welcome an issue or PR post merge if we want to address this.

Comment (Thels) [reply calm-widgets/src/widgets/list/list-template.html:1]: Given the urgency of needing this in - can you raise an issue for this and we can merge without?

File: calm-widgets/src/widget-registry.ts (hunk 4 above)
Comment (Thels) [reply calm-widgets/src/widget-registry.ts]: This should be handled by some automatic linting we run pre-PR.

This won't be the only case of this in the code-base.

File: calm-widgets/src/widget-helpers.spec.ts (hunk 3 above)
Comment (LeighFinegold) [reply calm-widgets/src/widget-helpers.spec.ts:130]: Yes the implementation handles that

File: calm-widgets/src/widget-registry.ts (hunk 4 above)
Comment (LeighFinegold) [reply calm-widgets/src/widget-registry.ts]: The ideal is the IDE would auto correct those. https://github.com/finos/architecture-as-code/issues/1396 has been raised to run strict mode. Raised https://github.com/finos/architecture-as-code/issues/1502. 

I can address this one instance, but will not fix the lint rule now as there are likely lots of other instances and occurances of hte same. This should be addressed in a future office hour.

Comment (LeighFinegold) [reply calm-widgets/src/widget-registry.ts]: I addressed the minor comment all the same on last commit

File: cli/src/cli.e2e.spec.ts (hunk 6 above)
Comment (LeighFinegold) [reply cli/src/cli.e2e.spec.ts]: I addressed the minor comment though on last commit

**Comment by rocketstack-matt** [comment]
> I think that this is a big PR and @LeighFinegold has put a considerable amount of his own personal time into getting this into the state it's in.
> 
> I've pulled down the PR and it's behaving, looks like @rocketstack-matt also has done it based on the conversations above. I think we should approve and then refactor/improve as we go on instead of trying to make it perfect from the get-go.
> 
> @markscott-ms - you good with this too?

Agreed - I've approved, @markscott-ms if you're happy with he issues being raised post merge lets move ahead.

**Comment by markscott-ms** [comment]
@Thels I agree with merging this in to allow divide and conquer on anything spotted.

Functionally and conceptually this looks great.

**Review by markscott-ms — APPROVED** [review]
